src/server/project.ts

```diff
@@ -1777,6 +1775,139 @@ namespace ts.server {
         }
     }
 
+    export class AutoImportProviderProject extends Project {
+        private static readonly newName = createProjectNameFactoryWithCounter(makeAutoImportProviderProjectName);
+
+        /*@internal*/
+        static getRootFileNames(dependencySelection: PackageJsonAutoImportPreference, hostProject: Project, compilerOptions: CompilerOptions): string[] {
+            if (!dependencySelection) {
+                return [];
+            }
+
+            let dependencyNames: Map<true> | undefined;
+            let rootNames: string[] | undefined;
+            const rootFileName = combinePaths(hostProject.currentDirectory, inferredTypesContainingFile);
+            const packageJsons = hostProject.getPackageJsonsForAutoImport(combinePaths(hostProject.currentDirectory, rootFileName));
+            for (const packageJson of packageJsons) {
+                packageJson.dependencies?.forEach((_, dependenyName) => addDependency(dependenyName));
+                packageJson.peerDependencies?.forEach((_, dependencyName) => addDependency(dependencyName));
+                if (dependencySelection === PackageJsonAutoImportPreference.All) {
+                    packageJson.devDependencies?.forEach((_, dependencyName) => addDependency(dependencyName));
+                }
+            }
+
+            if (dependencyNames) {

```

Comment:
Instead of doing this why not set options.types = dependencyNames here for this project so program constructution auto includes them ?

Reply:
You can modify program's getAutoTypeDirectiveNames to this instead..
```ts
// Thats correct solution project
const typeReferences: string[] = !rootNames.length && projectReferences ?  emptyArray : getAutomaticTypeDirectiveNames(options, host) ;
```

Reply:
Oh, I had to use this approach back when I was trying a single program for all projects’ dependencies because there was no suitable `currentDirectory` that could resolve types in sibling projects with their own node_modules. But now that there’s one program per project, I think this will work.

Reply:
Well... I think I would still need to resolve them here, because I need to check if the main program already contains that dependency, which is really easy when you have a resolved fileName (this happens just a few lines below your comment). Is there a better way to do that?

Reply:
On AutoImportProvider Projhect override resolveTypeReferenceDirective to do `super.` thing and then if file is in main project return unresolved result otherwise return actual result?

Reply:
You want to do it only for inferredTypeContainingFile though... if depA has typeReferenceDirective to depB and depB is already in project, you still want depA in the auto import provider to get correct types for depA as it probably depends on it

Reply:
The reason I initially stopped using `Project.resolveTypeReferenceDirective` was because it created a weird watch state via the resolution cache, which ended up breaking things. When I switched to these utility methods it only slowed down by a couple ms per package.json and the watch issues were fixed. This approach seems to be working well, so I’m hesitant to reintroduce that complexity.

Reply:
I think it was because you were watching at wrong time.. Infact resolutionCache is suppose to handle the caching/updates so i think its worth the try for cleaner approach. I dont think we should use main project to do resolutions for this as it may have some additional logics on when and what to resolve or which files to assume as present etc and that shouldnt play role into it.. If it shoiuld, it needs to be more explicit

Reply:
The next problem with this is that it means we can’t skip creating the AutoImportProviderProject if the main program already contains the dependencies, because we need a project instance to call `resolveTypeReferenceDirectives` from, but if it ends up returning an array of all `undefined` because the main program contained everything, we just wasted a lot of allocations creating the project in the first place 😕